### DISCUSSION TITLE
Automatic loading of remote branches

### DISCUSSION THREAD

**Discussion by meruiden** [ProjectBorealis/UEGitPlugin#177] [Show and tell]
Hi,

For a project I'm working on, we work with feature branches. I need the plugin to automatically load the branches, for example by using ```git branch -r``` and add them to the RegisterStateBranches list. Right now it's rather frustrating to have to add the branches, recompile and restart the whole time.

I'd like to know your thoughts on this as well. Thanks!

**Comment by mastercoms** [comment]
Hi, you're free to do any query for branches within your own code using the state branches API. I'd be happy to also take a PR adding this as an option, but I'd be interested in knowing how we could generically resolve state branch order on our own. Perhaps we need 2 concepts within the plugin, one for state branches and another for feature branches (orderless, always higher order than current branch).

**Comment by meruiden** [comment]
Hi,

For us the only thing that really matters is that you can't checkout files that have been modified on other feature branches until you get the changes into the current branch. Right now the ordering for this doesn't really matter right?

As for querying the branches, is there a build in function for querying remote branches :o ? 

Thanks!

**Comment by mastercoms** [comment]
You can use `RunCommand` in `GitSourceControlUtils` to do this.

**Comment by meruiden** [comment]
I noticed the function, but the library is currently setup so that I can't include the GitSourceControlUtils from my editor module, I get linker errors when I try to call any of the plugin's functions, even though I have added it to ```PrivateDependencyModuleNames```

**Comment by mastercoms** [comment]
No worries, I'll look into providing something easier soon.

**Comment by meruiden** [comment]
Thanks! I think the linker error solution is to add ```GITSOURCECONTROL_API``` to the parts that need exposing. I'll use that to fix my issues for now. I'll make sure to check for updates 👀

**Comment by meruiden** [comment]
Solved it for now by adding the ```GITSOURCECONTROL_API``` and running the below code every 3 seconds:

```cpp
Async(EAsyncExecution::Thread, [ = ] {
    FGitSourceControlModule & GitSourceControl = FGitSourceControlModule::Get();

    if (!GitSourceControl.GetProvider().IsGitAvailable()) {
        return;
    }

    const FString & PathToGitBinary = GitSourceControl.AccessSettings().GetBinaryPath();
    const FString & RepositoryRoot = GitSourceControl.GetProvider().GetPathToRepositoryRoot();

    TArray < FString > Results;
    TArray < FString > Errors;
    GitSourceControlUtils::RunCommand("branch", PathToGitBinary, RepositoryRoot, TArray < FString > ({
        "-r"
    }), TArray < FString > (), Results, Errors);

    TArray < FString > Branches;
    const TArray < FString > CoreBranches = {
        "origin/main",
        "origin/beta",
        "origin/production"
    };

    for (int i = 0; i < Results.Num(); i++) {
        FString Result = Results[i];

        if (Result.Contains("->")) {
            TArray < FString > Splitted;
            Result.ParseIntoArray(Splitted, TEXT("->"), true);
            if (Splitted.Num() == 2) {
                Result = Splitted[1];
            }
        }

        Result.RemoveSpacesInline();

        if (!CoreBranches.Contains(Result)) {
            Branches.Add(Result);
        }
    }

    Branches.Append(CoreBranches);

    Async(EAsyncExecution::TaskGraphMainThread, [Branches] {
        const ISourceControlModule & SourceControlModule = ISourceControlModule::Get();
        ISourceControlProvider & SourceControlProvider = SourceControlModule.GetProvider();
        SourceControlProvider.RegisterStateBranches(Branches, TEXT("Content"));
    });
});
```

**Comment by mastercoms** [comment]
Fixed in https://github.com/ProjectBorealis/UEGitPlugin/commit/cf8a04f86f273f697d45e0d1a7f37fc60a02630f
